File: packages/core/src/engine/verbs/recode.ts
/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import { escape, op } from 'arquero'
import ColumnTable from 'arquero/dist/types/table/column-table'
import { ExprObject } from 'arquero/dist/types/table/transformable'
import { TableStore } from '../..'
import { RecodeArgs, Step } from '../../types'

/**
 * Executes an arquero derive to map a list of values to new values.
 * Commonly used for recategorization.
 * @param step
 * @param store
 * @returns
 */
export async function recode(
	step: Step,
	store: TableStore,
): Promise<ColumnTable> {
	const { input, args } = step
	const { column, as, map } = args as RecodeArgs
	const inputTable = await store.get(input)

	const dArgs: ExprObject = {
		[as]: escape((d: any) => op.recode(d[column], map)),
	}

	return inputTable.derive(dArgs)
}

File: python/data-wrangling-components/data_wrangling_components/engine/verbs/recode.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project.
#

from typing import Dict

from data_wrangling_components.table_store import TableContainer


class RecodeMap(dict):
    def __missing__(self, key):
        return key


def recode(input: TableContainer, to: str, column: str, map: Dict):
    map = RecodeMap(map)

    input_table = input.table

    output = input_table.copy()
    output[to] = output[column].map(map)
    return TableContainer(table=output)
